src/main/java/com/cleanroommc/groovyscript/helper/ingredient/IngredientBase.java

```diff
@@ -1,49 +1,80 @@
 package com.cleanroommc.groovyscript.helper.ingredient;
 
 import com.cleanroommc.groovyscript.api.IIngredient;
-import com.cleanroommc.groovyscript.sandbox.ClosureHelper;
-import groovy.lang.Closure;
+import com.cleanroommc.groovyscript.compat.vanilla.ItemStackMixinExpansion;
+import com.cleanroommc.groovyscript.compat.vanilla.ItemStackTransformer;
 import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
 import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.event.ForgeEventFactory;
 import org.jetbrains.annotations.Nullable;
 
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+
 public abstract class IngredientBase implements IIngredient {
 
-    protected Closure<Object> matchCondition;
-    protected Closure<Object> transformer;
+    protected Predicate<ItemStack> matchCondition;
+    protected ItemStackTransformer transformer;
     protected String mark;
 
-    public IngredientBase when(Closure<Object> matchCondition) {
+    public IngredientBase when(Predicate<ItemStack> matchCondition) {
         IngredientBase fresh = (IngredientBase) this.exactCopy();
         fresh.matchCondition = matchCondition;
         return fresh;
     }
 
-    public IngredientBase transform(Closure<Object> transformer) {
+    public IngredientBase transform(ItemStackTransformer transformer) {
         IngredientBase fresh = (IngredientBase) this.exactCopy();
         fresh.transformer = transformer;
         return fresh;
     }
 
+    public IngredientBase transformDamage(int amount) {
+        // reliably set itemDamage field of item stack
+        return transform(self -> IngredientHelper.damageItem(self, amount));
+    }
+
+    public IngredientBase transformDamage() {
+        return transformDamage(1);
+    }
+
+    public IngredientBase transformNbt(UnaryOperator<NBTTagCompound> transformer) {
+        return transform(self -> {
+            ItemStackMixinExpansion.of(self).exactCopy().grs$getItemStack().setTagCompound(transformer.apply(self.getTagCompound()));
+            return self;
+        });
+    }
+
     public IngredientBase reuse() {
-        return transform(IngredientHelper.REUSE);
+        return transform(self -> self);
+    }
+
+    public IngredientBase noreturn() {
```

Comment:
should this be `noReturn`? if so, should probably keep `noreturn` and either have it be a duplicate or deprecate it.